src/Viewer.tsx

```diff
@@ -853,6 +855,44 @@ function Viewer(): ReactElement {
     }
   }
 
+  // TODO: Move to a separate component?
+  const hoverTooltipContent = [
+    <p key="track_id">Track ID: {lastHoveredId && dataset?.getTrackId(lastHoveredId)}</p>,
+    <p key="feature_value">
+      {dataset?.getFeatureName(featureKey) || "Feature"}:{" "}
+      <span style={{ whiteSpace: "nowrap" }}>{hoveredFeatureValue}</span>
+    </p>,
+  ];
+
+  if (config.vectorConfig.visible && lastHoveredId !== null && motionDeltas) {
+    const vectorKey = config.vectorConfig.key;
+    const vectorName = vectorKey === VECTOR_KEY_MOTION_DELTA ? "Avg. motion delta" : vectorKey;
+    const motionDelta = [motionDeltas[2 * lastHoveredId], motionDeltas[2 * lastHoveredId + 1]];
+    if (Number.isNaN(motionDelta[0]) || Number.isNaN(motionDelta[1])) {
+      // {vector}:
+      hoverTooltipContent.push(<p key="vector_empty">{vectorName}:</p>);
+    } else if (config.vectorConfig.tooltipMode === VectorTooltipMode.MAGNITUDE) {
+      const magnitude = Math.sqrt(motionDelta[0] ** 2 + motionDelta[1] ** 2);
+      const angleDegrees = (360 + Math.atan2(-motionDelta[1], motionDelta[0]) * (180 / Math.PI)) % 360;
+
+      // {vector}: {magnitude} {unit}, {angle}°
+      hoverTooltipContent.push(
+        <p key="vector_magnitude">
+          {vectorName}: {numberToStringDecimal(magnitude, 3)} px, {numberToStringDecimal(angleDegrees, 1)}°
+        </p>
+      );
+    } else {
+      // {vector}: ({x}, {y}) {unit}
+      const allowIntegerTruncation = Number.isInteger(motionDelta[0]) && Number.isInteger(motionDelta[1]);
+      hoverTooltipContent.push(
+        <p key="vector_components">
+          {vectorName}: ({numberToStringDecimal(motionDelta[0], 3, allowIntegerTruncation)},{" "}
+          {numberToStringDecimal(motionDelta[1], 3, allowIntegerTruncation)}) px
+        </p>
```

Comment:
I don't typically like to have this much data grooming mixed up with rendering. I think this would be a lot easier to follow if you had a separate function that checked for and got the appropriate data and another one that created the <p> and pushed it to the array.

Reply:
Done!